### DISCUSSION TITLE
No system prompt for Anthropic?

### DISCUSSION THREAD

**Discussion by eterps** [frankroeder/parrot.nvim#22] [Q&A] [answered]
Am I right in assuming that currently there is no support for `system_prompt` for Antropic provider?

https://github.com/frankroeder/parrot.nvim/blob/2719099858898447dfb3f74b2af8ef7b8bd388c3/lua/parrot/provider/anthropic.lua#L50

**Comment by eterps** [comment]
Hmmm, it seems it should support `system_prompt`, however I doubt if it works.
What is the best way to check this?
I see there are logger statements, do I need to activate logging somehow? And where does it go?

Maybe obivous questions for people who have authored nvim plugins more often, but I'm a bit in the dark on this.

**Comment by frankroeder** [comment, marked as answer]
Anthropic expects the system prompt inside the curl parameters (https://docs.anthropic.com/en/api/messages) and not inside the messages with `role = system`, like the OpenAI API.

Yes, it definitely would be helpful to write a comment or document it somewhere.
By default, all providers have the system prompt defines as OpenAI. However, for Anthropic I implement this exception by removing the default system prompt https://github.com/frankroeder/parrot.nvim/blob/2719099858898447dfb3f74b2af8ef7b8bd388c3/lua/parrot/provider/anthropic.lua#L43 and adding it later to the curl call via the agent.model https://github.com/frankroeder/parrot.nvim/blob/2719099858898447dfb3f74b2af8ef7b8bd388c3/lua/parrot/agents.lua#L119

You can see my print statements for the messages, header and agent.model below, comparing Anthropic and OpenAI: 
<img width="1265" alt="screenshot 2024-07-02 at 09 21 46" src="https://github.com/frankroeder/parrot.nvim/assets/19746932/9ed57073-a80d-4594-aaa9-d1565c18bb44">

I hope this helps.

**Comment by eterps** [comment]
Thanks for the explanation @frankroeder , it seems the system prompt doesn't end up in the curl params.
I'll see if I can find out why. These are the params in my case:

```
parrot.nvim: debug_curl_params: { "--no-buffer", "--silent", "-H", "accept: application/json", "-H", "content-type: application/json", "-d", '{"messages":[{"content":"2+2=?","role":"user"}],"model":"claude-3-5-sonnet-20240620", "max_tokens":4096,"stream":true}', "https://api.anthropic.com/v1/messages", "-H", "x-api-key: xxx", "-H", "anthropic-version: 2023-06-01" }
```

**Comment by frankroeder** [comment]
Hi @eterps, they definitely end up in the request, otherwise the topic model would not be able to generate a 2-3 word topic headlines for the chats. Have a look at the `prepare_payload` utils function and the output it generates. Search for it in the init.lua file.
<img width="2541" alt="screenshot 2024-07-02 at 11 12 53" src="https://github.com/frankroeder/parrot.nvim/assets/19746932/91834997-a488-44c8-94cd-bd9d87d0250a">

BTW, I think it is always good to double proof things. I really appreciate your effort!

**Comment by eterps** [comment]
Ok, found some more details.
The system prompt does not always end up in the curl request.

For instance, this config results in the curl params below it:

```lua
{
  "frankroeder/parrot.nvim",
  dependencies = { 'ibhagwan/fzf-lua', 'nvim-lua/plenary.nvim' },
  opts = {
    providers = {
      anthropic = { api_key = os.getenv "ANTHROPIC_API_KEY" }
    }
  }
}
```

```
parrot.nvim: curl_params: { "--no-buffer", "--silent", "-H", "accept: application/json", "-H", "content-type: application/json", "-d", "{\"max_tokens\":4096,\"stream\":true,\"model\":\"claude-3-haiku-20240307\",\"system\":\"You are a versatile AI assistant with capabilities\\nextending to general knowledge and coding support. When engaging\\nwith users, please adhere to the following guidelines to ensure\\nthe highest quality of interaction:\\n\\n- Admit when unsure by saying 'I don't know.'\\n- Ask for clarification when needed.\\n- Use first principles thinking to analyze queries.\\n- Start with the big picture, then focus on details.\\n- Apply the Socratic method to enhance understanding.\\n- Include all necessary code in your responses.\\n- Stay calm and confident with each task.\\n\",\"messages\":[{\"role\":\"user\",\"content\":\"Hi\"}]}", "https://api.anthropic.com/v1/messages", "-H", "x-api-key: xxx", "-H", "anthropic-version: 2023-06-01"}
```

```
parrot.nvim: curl_params: { "--no-buffer", "--silent", "-H", "accept: application/json", "-H", "content-type: application/json", "-d", "{\"max_tokens\":32,\"stream\":true,\"model\":\"claude-3-sonnet-20240229\",\"system\":\"Summarize the topic of our conversation above\\nin two or three words. Respond only with those words.\\n\",\"messages\":[{\"role\":\"user\",\"content\":\"Hi\"},{\"role\":\"assistant\",\"content\":\"Hello! How can I assist you today? I'm here to help with any questions or tasks you may have. Please feel free to provide me with more details about what you need, and I'll do my best to provide a helpful and thoughtful response.\"},{\"role\":\"user\",\"content\":\"You only respond with 2 to 3 words to summarize the past conversation.\"}]}", "https://api.anthropic.com/v1/messages", "-H", "x-api-key: xxx", "-H", "anthropic-version: 2023-06-01" }
```

While this config results in the curl params below (notice the system prompt missing in the first, but still present in the second curl request):

```lua
{
  "frankroeder/parrot.nvim",
  dependencies = { 'ibhagwan/fzf-lua', 'nvim-lua/plenary.nvim' },
  opts = {
    providers = {
      anthropic = { api_key = os.getenv "ANTHROPIC_API_KEY" }
    }
    agents = {
      chat = {
        {
          name = "CustomClaude",
          provider = "anthropic",
          model = {
            model = "claude-3-5-sonnet-20240620",
            max_tokens = 4096
          },
          system_prompt = "Help me!"
        }
      }
    }
  }
}
```

```
parrot.nvim: curl_params: { "--no-buffer", "--silent", "-H", "accept: application/json", "-H", "content-type: application/json", "-d", '{"max_tokens":4096,"messages":[{"content":"Hi","role":"user"}],"model":"claude-3-5-sonnet-20240620","stream":true}', "https://api.anthropic.com/v1/messages", "-H", "x-api-key: xxx", "-H", "anthropic-version: 2023-06-01" }
```

```
parrot.nvim: curl_params: { "--no-buffer", "--silent", "-H", "accept: application/json", "-H", "content-type: application/json", "-d", "{\"max_tokens\":32,\"messages\":[{\"content\":\"Hi\",\"role\":\"user\"},{\"content\":\"Hello! How can I assist you today? Feel free to ask any questions or let me know if there's anything you'd like to discuss.\",\"role\":\"assistant\"},{\"content\":\"You only respond with 2 to 3 words to summarize the past conversation.\",\"role\":\"user\"}],\"stream\":true,\"model\":\"claude-3-sonnet-20240229\",\"system\":\"Summarize the topic of our conversation above\\nin two or three words. Respond only with those words.\\n\"}", "https://api.anthropic.com/v1/messages", "-H", "x-api-key: xxx", "-H", "anthropic-version: 2023-06-01" }
```

I'll see if I can locate the problem, but if you have an idea what might be wrong I would appreciate it.

**Reply by frankroeder** [reply]
From what I can see, you called it "system_prompt", while it should be called "system" as you can see in the documentation linked above. In addition, you need to put `"system" = "Help me!"` inside the model list.

```lua
chat = {
        {
          name = "CustomClaude",
          provider = "anthropic",
          model = {
            model = "claude-3-5-sonnet-20240620",
            max_tokens = 4096,
            system = "Help me!" -- is passed to curl
          },
          system_prompt = "" -- ignored for anthropic
        }
      }
```

**Reply by eterps** [reply]
Thanks, that helps clarify things a bit. However, when using your example config, I'm getting the following error:

```
parrot.nvim/lua/parrot/init.lua:105: table index is nil
```

I'll have to leave this for now, but maybe I can look into it further during the week.

I initially misunderstood how parrot.nvim is configured, thinking it was an abstraction for all providers and that the provider-specific details were handled internally. Your explanations have helped me understand it better.

I'm also wondering how Neovim plugins in general validate user configurations. Doing it ad-hoc within the code seems rather tedious and complex. I just found this library: https://github.com/sschoener/lua-schema, which seems more suitable for this purpose. Using a schema library could also make the configuration handling code much simpler, since it would know the provided config is correct and wouldn't need to do additional checks, error handling, and user logging.

Additionally, user config could be merged onto default config using [vim.tbl_deep_extend()](https://neovim.io/doc/user/lua.html#vim.tbl_deep_extend())."

**Reply by frankroeder** [reply]
Indeed, it will require some time to reorganize the entire repository and make these processes more intuitive. The next logical action would be to ensure the provider/agent configuration is consistent, and internally manage differences, as those with Anthropic.

**Reply by eterps** [reply]
> From what I can see, you called it "system_prompt", while it should be called "system" as you can see in the documentation linked above. In addition, you need to put `"system" = "Help me!"` inside the model list.

Thanks, I eventually got it working.

I also found out that hooks need to be configured like this for Anthropic LLMs (putting it here for reference):

```lua
        ProofReader = function(prt, params)
          local chat_system_prompt = "I want you to act as a proofreader."
          local agent = prt.get_chat_agent()
          agent.model.system = chat_system_prompt
          prt.cmd.ChatNew(params, agent.model, '')
        end,
```

**Reply by frankroeder** [reply]
Yes, forget about this hook. I want to structure it differently. This is actually deprecated. Other hooks use templates and those end up within the usual messages.
Thank you for pointing this out.

**Reply by eterps** [reply]
@frankroeder I'm still quite new to Lua and Neovim plugin development, but could the configuration handling be simplifed by using something like this?

- It does a full merge of the user configuration into a copy of the default configuration
- It merges the provider defaults of those providers the user actually configured
- It merges the agents for the providers the user has configured

I was curious how that could be solved, so I came up with this.
(`user_config` is just an example here)

```lua
local default_config = require("parrot.config")

local user_config = {
  chat_confirm_delete = false,
  chat_dir = '~/.chats',
  providers = {
    anthropic = {
      api_key = os.getenv "ANTHROPIC_API_KEY",
    }
  },
  agents = {
    chat = {
      {
        name = "Claude-3.5-Sonnet",
        provider = "anthropic",
        model = {
          model = "claude-3-5-sonnet-20240620",
          max_tokens = 4096,
          system = [[You are a versatile AI assistant]],
        },
        system_prompt = ""
      }
    }
  },
  hooks = {
    --PrtProofReader
    ProofReader = function(prt, params)
      local chat_system_prompt = [[I want you to act as a proofreader.]]
      local agent = prt.get_chat_agent()
      agent.model.system = chat_system_prompt
      prt.cmd.ChatNew(params, agent.model, '')
    end
  }
}

local function merge_providers(default_providers, user_providers)
  local result = {}
  for provider, config in pairs(user_providers) do
    result[provider] = vim.tbl_deep_extend('force', default_providers[provider] or {}, config)
  end
  return result
end

local function merge_agent_type(default_agents, user_agents, user_providers)
  local merged = vim.deepcopy(user_agents) or {}
  for _, default_agent in ipairs(default_agents) do
    if user_providers[default_agent.provider] then
      table.insert(merged, vim.deepcopy(default_agent))
    end
  end
  return merged
end

local function merge_agents(default_agents, user_agents, user_providers)
  return {
    command = merge_agent_type(default_agents.command or {}, user_agents.command, user_providers),
    chat = merge_agent_type(default_agents.chat or {}, user_agents.chat, user_providers)
  }
end

local config = vim.tbl_deep_extend('force', default_config, user_config)

config.providers = merge_providers(default_config.providers, user_config.providers)
config.agents = merge_agents(default_config.agents or {}, user_config.agents or {}, config.providers)

print(vim.inspect(config))
```

**Reply by frankroeder** [reply]
@eterps Looks good! Feel free to create a pull request that we both can work on this.

**Reply by frankroeder** [reply]
I already created a local branch and there is currently no merge of the agents happening.
```lua
local merged = vim.deepcopy(user_agents) or {}
```
You either go with the default agent or the user agent.

Feel free to contribute: https://github.com/frankroeder/parrot.nvim/pull/23

**Reply by eterps** [reply]
> You either go with the default agent or the user agent.

Thanks, I can fix that.

**Reply by eterps** [reply]
@frankroeder, I need clarification on the provider configuration behavior. When a user configures the providers section of their configuration as follows:

```lua
providers = {
    anthropic = { api_key = os.getenv "ANTHROPIC_API_KEY" },
    openai = { api_key = os.getenv "OPENAI_API_KEY" }
 },
 ```
 
The intuitive expectation would be that only agents from these two specified providers (Anthropic and OpenAI) would be available for use. This would imply that the merging of providers only combines the default configurations for `anthropic` and `openai`, while excluding all other providers.

Would you agree that this is the behaviour a user would expect? Or did you have another mechanism in mind to 'not use' certain providers?

**Reply by frankroeder** [reply]
Hi, I am already working on this on a different local branch, only making the providers available the user configures correctly, hence providing the API key or something along that lines.

**Reply by eterps** [reply]
> Hi, I am already working on this on a different local branch, only making the providers available the user configures correctly, hence providing the API key or something along that lines.

Yes, I am also on that branch, that's why I am asking, because this loop also merges providers the user didn't configure:

https://github.com/frankroeder/parrot.nvim/pull/23/files#diff-8f49ff122fed688344cbdc1791680b557dabd81af784a9afa96925fa297e22c0R52

I would assume that if the table `providers` omits certain keys (like `pplx` or `mistral` for example) those defaults shouldn't be copied in. But currently it does.

> only making the providers available the user configures correctly

Whether a configuration is 'correct' is a separate issue I would think. An 'absence' of a provider is not invalid IMO.

But maybe I am misunderstanding things, I am just happy to help out if possible.

**Reply by frankroeder** [reply]
That is actually a good idea. Yes, my bad. Validation should be happening at a different stage. Let us take this discussion to the pull request and feel free to push your changes.

**Reply by eterps** [reply]
Cool, thanks. As a result, I believe it will also allow the code to be much simpler in the end. Let's take this slowly to come up with a good solution. I will continue the discussion in the pull request from now on.
